packages/babel-plugin-transform-typescript/src/namespace.js

```diff
@@ -0,0 +1,205 @@
+  TSModuleDeclaration:
+    "A namespace may not share the name of its parent namespace.",
+};
+
+export default function transpileNamespace(path, t) {
+  if (path.node.declare || path.node.id.type === "StringLiteral") {
+    path.remove();
+    return;
+  }
+
+  const name = path.node.id.name;
+  const value = handleNested(path, t, JSON.parse(JSON.stringify(path.node)));
+  const bound = path.scope.hasOwnBinding(name);
+  if (path.parent.type === "ExportNamedDeclaration") {
+    if (!bound) {
+      path.parentPath.insertAfter(value);
+      path.replaceWith(getDeclaration(t, name));
+      path.scope.registerDeclaration(path.parentPath);
+    } else {
+      path.parentPath.replaceWith(value);
+    }
+  } else if (bound) {
+    path.replaceWith(value);
+  } else {
+    path.scope.registerDeclaration(
+      path.replaceWithMultiple([getDeclaration(t, name), value])[0],
+    );
+  }
+}
+
+function getDeclaration(t, name) {
+  return t.variableDeclaration("let", [
+    t.variableDeclarator(t.identifier(name)),
+  ]);
+}
+
+function getMemberExpression(t, name, itemName) {
+  return t.memberExpression(t.identifier(name), t.identifier(itemName));
+}
+
+function handleNested(path, t, node, parentExportName) {
+  const names = [];
+  const name = node.id.name;
+  const namespaceTopLevel = node.body.body;
+  for (let i = 0; i < namespaceTopLevel.length; i++) {
+    const subNode = namespaceTopLevel[i];
+    switch (subNode.type) {
+      case "TSModuleDeclaration": {
+        const moduleName = subNode.id.name;
+        if (moduleName === name) {
+          throw path.hub.file.buildCodeFrameError(
+            subNode,
+            SHARED_NAMESPACE_ERRORS.TSModuleDeclaration,
+          );
+        }
+        if (names[moduleName]) {
+          namespaceTopLevel[i] = handleNested(path, t, subNode);
+        } else {
+          names[moduleName] = true;
+          namespaceTopLevel.splice(
+            i++,
+            1,
+            getDeclaration(t, moduleName),
+            handleNested(path, t, subNode),
+          );
+        }
+        continue;
+      }
+      case "TSEnumDeclaration":
+      case "FunctionDeclaration":
+      case "ClassDeclaration": {
+        const itemName = subNode.id.name;
+        if (itemName === name) {
+          throw path.hub.file.buildCodeFrameError(
+            subNode,
+            SHARED_NAMESPACE_ERRORS[subNode.type],
+          );
+        }
+        names[itemName] = true;
+        continue;
+      }
+      case "VariableDeclaration":
+        for (const variable of subNode.declarations) {
+          const variableName = variable.id.name;
+          if (variableName === name) {
+            throw path.hub.file.buildCodeFrameError(
+              variable,
+              SHARED_NAMESPACE_ERRORS.VariableDeclaration,
+            );
+          }
+          names[variableName] = true;
+        }
+        continue;
+      default:
+        continue;
+      case "ExportNamedDeclaration":
+    }
```

Comment:
Nodes that are exporting (`ExportNamedDeclaration`) need to get transpiled. The rest/other switch of the method deals with those cases, which are effectively the *primary case* (what really matters) for the entire function. The other cases are simplistically tracking the usages of names, effectively incidental.

Alternatives:

* I can move the entire switch statement up into that case and drop the default-case. I didn't do this because it felt unnecessary.

* I can add a comment(s) explaining that the case that matters most is just the `export` case, and anything else just continues in the loop.

* I can the second switch into a function, to make it more concise.  I didn't do this because it's a lot of parameter sharing, especially for a *recursive* function. However, I would prefer this over *nesting* switches (first alternative).

Reply:
> I can add a comment(s) explaining that the case that matters most is just the export case, and anything else just continues in the loop.

:+1: